// @babel/plugin-transform-block-scoping/lib/tdz.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.visitor = void 0;

var _core = require("@babel/core");

function getTDZStatus(refPath, bindingPath) {
  const executionStatus = bindingPath._guessExecutionStatusRelativeTo(refPath);

  if (executionStatus === "before") {
    return "outside";
  } else if (executionStatus === "after") {
    return "inside";
  } else {
    return "maybe";
  }
}

function buildTDZAssert(node, state) {
  return _core.types.callExpression(state.addHelper("temporalRef"), [node, _core.types.stringLiteral(node.name)]);
}

function isReference(node, scope, state) {
  const declared = state.letReferences.get(node.name);
  if (!declared) return false;
  return scope.getBindingIdentifier(node.name) === declared;
}

const visitedMaybeTDZNodes = new WeakSet();
const visitor = {
  ReferencedIdentifier(path, state) {
    if (!state.tdzEnabled) return;
    const {
      node,
      parent,
      scope
    } = path;

    // Return early if the node is in a for loop
    if (path.parentPath.isFor({
      left: node
    }}) return;

    // Return early if the node is not a reference
    if (!isReference(node, scope, state)) return;

    const bindingPath = scope.getBinding(node.name).path;

    // Return early if the binding is a function declaration
    if (bindingPath.isFunctionDeclaration()) return;

    const status = getTDZStatus(path, bindingPath);

    // Return early if the node is outside the temporal dead zone
    if (status === "outside") return;

    if (status === "maybe") {
      if (visitedMaybeTDZNodes.has(node)) return;

      visitedMaybeTDZNodes.add(node);
      const assert = buildTDZAssert(node, state);
      bindingPath.parent._tdzThis = true;

      if (path.parentPath.isUpdateExpression()) {
        if (parent._ignoreBlockScopingTDZ) return;
        path.parentPath.replaceWith(_core.types.sequenceExpression([assert, parent]));
      } else {
        path.replaceWith(_core.types.expressionStatement(assert));
      }
    } else if (status === "inside") {
      path.replaceWith(_core.types.expressionStatement(_core.template.ast`${state.addHelper("tdz")}("${node.name}")`));
    }
  },

  AssignmentExpression: {
    exit(path, state) {
      if (!state.tdzEnabled) return;
      const {
        node
      } = path;

      // Return early if the node ignores block scoping TDZ
      if (node._ignoreBlockScopingTDZ) return;

      const nodes = [];
      const ids = path.getBindingIdentifiers();

      Object.entries(ids).forEach(([name, id]) => {
        // Return early if the node is not a reference
        if (!isReference(id, path.scope, state)) return;

        nodes.push(id);
      });

      // Return early if there are no nodes to process
      if (nodes.length === 0) return;

      node._ignoreBlockScopingTDZ = true;
      nodes.push(node);
      path.replaceWithMultiple(nodes.map(n => _core.types.expressionStatement(n)));
    }

  }
};
exports.visitor = visitor;
