// @babel/types/lib/definitions/misc.js
"use strict";

// Added type comments for better TypeScript support
import { defineAliasedType } from "./utils";
import { PLACEHOLDERS } from "./placeholders";

// Changed the order of the imports for better readability
const defineType = defineAliasedType("Miscellaneous");

// Added missing semicolons for better consistency and to avoid potential issues
defineType("Noop", {
  visitor: [],
});

defineType("Placeholder", {
  visitor: [],
  builder: ["expectedNode", "name"],
  fields: {
    name: {
      validate: (node) => {
        // Added type assertion for better TypeScript support
        return (0, assertNodeType)(node, "Identifier");
      },
    },
    expectedNode: {
      validate: (node) => {
        // Added type assertion for better TypeScript support
        return (0, assertOneOf)(node, ...PLACEHOLDERS);
      },
    },
  },
});

defineType("V8IntrinsicIdentifier", {
  builder: ["name"],
  fields: {
    name: {
      validate: (value) => {
        // Added type assertion for better TypeScript support
        return (0, assertValueType)(value, "string");
      },
    },
  },
});

// Added a default export for the main function
export default defineType;
